utils/senslog.c

```diff
@@ -0,0 +1,249 @@
+			case 'm':
+				senslog_common.flags.mag = true;
+				break;
+
+			case 'b':
+				senslog_common.flags.baro = true;
+				break;
+
+			default:
+				fprintf(stderr, "Unknown sensor log option: '%c'\n", *devs);
+				err = true;
+				break;
+		}
+
+		devs++;
+	}
+
+	/* At least one initialized device and no error means successful parsing */
+	if ((senslog_common.flags.accel || senslog_common.flags.gyro || senslog_common.flags.mag || senslog_common.flags.baro) && !err) {
+		return 0;
+	}
+
+	return -1;
+}
+
+
+/* 
+* Parsing arguments. Returns:
+* 0 if all variables were successfully initialized, and application is good to go.
+* -1 if an error occurred during parsing
+* +1 if only help was printed and application shall exit.
+ */
+static int senslog_parseArgs(int argc, char **argv)
+{
+	int c;
+	const char *devs = NULL;
+
+	senslog_common.step = 0;
+	senslog_common.diffMax = 0;
+	senslog_common.raw = false;
+
+	if (argc < 2) {
+		senslog_help(argv[0]);
+		return -1;
+	}
+
+	while ((c = getopt(argc, argv, "ht:s:d:r")) != -1) {
+		switch (c) {
+			case 'h':
+				senslog_help(argv[0]);
+				return +1; /* only help message printed, returning +1 */
+
+			case 't': /* time set */
+				senslog_common.diffMax = atoi(optarg) * 1000000;
+				break;
+
+			case 's': /* samppling interval set */
+				senslog_common.step = atoi(optarg) * 1000;
+				break;
+
+			case 'd': /* devices to log */
+				devs = optarg;
+				break;
+
+			case 'r': /* do not use corrections, raw mode */
+				senslog_common.raw = true;
+				break;
+
+			default: /* any other case is an error */
+				senslog_help(argv[0]);
+				return -1;
+		}
+	}
+
+	if (senslog_common.step == 0 || senslog_common.diffMax == 0 || senslog_parseDevices(devs) < 0) {
+		fprintf(stderr, "Invalid arguments!\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+int main(int argc, char **argv)
+{
+	FILE *file;
+	sensor_event_t accelEvt, gyroEvt, magEvt, baroEvt;
+	char buf[128];
+	time_t t0, now, diff;
+	int res, cnt;
+
+	res = senslog_parseArgs(argc, argv);
+	if (res != 0) {
+		return (res < 0) ? EXIT_FAILURE : EXIT_SUCCESS;
+	}
+
+	fprintf(stdout, "Logging for %lli seconds with %llims step\n", senslog_common.diffMax / 1000000, senslog_common.step / 1000);
+
+	/* initialization of sensor client with selectable corrections (raw == true means corrections disabled) */
+	if (sensc_init("/dev/sensors", !senslog_common.raw) < 0) {
+		fprintf(stderr, "Cannot initialize sensor client\n");
+		return EXIT_FAILURE;
+	}
+
+	file = fopen("/etc/senslog.txt", "w");
+	if (file == NULL) {
+		sensc_deinit();
+		fprintf(stderr, "Cannot open '/etc/altlog.txt' logfile\n");
+		return EXIT_FAILURE;
+	}
+
+	gettime(&t0, NULL);
+	now = t0;
+	diff = 0;
+
+	/* Main logging loop */
+	while (diff < senslog_common.diffMax) {
+		gettime(&now, NULL);
+		diff = now - t0;
+
+		if (senslog_common.flags.accel || senslog_common.flags.gyro || senslog_common.flags.mag) {
+			sensc_imuGet(&accelEvt, &gyroEvt, &magEvt);
+		}
+		if (senslog_common.flags.baro) {
```

Comment:
There are separate conditions:
```suggestion
		}
		
		if (senslog_common.flags.baro) {
```